src/openrct2-ui/windows/ServerList.cpp

```diff
@@ -73,529 +71,519 @@ static Widget window_server_list_widgets[] = {
+        auto beginBracketIndex = address.find('[');
+        auto endBracketIndex = address.find(']');
+        auto dotIndex = address.find('.');
+        auto colonIndex = address.find_last_of(':');
+        if (colonIndex != std::string::npos)
         {
+            if (endBracketIndex != std::string::npos || dotIndex != std::string::npos)
+            {
+                auto ret = std::sscanf(&address[colonIndex + 1], "%d", &port);
+                assert(ret);
+                if (ret > 0)
+                {
+                    address = address.substr(0, colonIndex);
+                }
+            }
         }
 
-        auto status = STR_NONE;
-        try
+        if (beginBracketIndex != std::string::npos && endBracketIndex != std::string::npos)
         {
-            auto entries = wanF.get();
-            allEntries.insert(allEntries.end(), entries.begin(), entries.end());
+            address = address.substr(beginBracketIndex + 1, endBracketIndex - beginBracketIndex - 1);
         }
-        catch (const MasterServerException& e)
-        {
-            status = e.StatusText;
-        }
-        catch (...)
+
+        if (!NetworkBeginClient(address, port))
         {
-            status = STR_SERVER_LIST_NO_CONNECTION;
+            ContextShowError(STR_UNABLE_TO_CONNECT_TO_SERVER, STR_NONE, {});
         }
-        return std::make_tuple(allEntries, status);
-    });
-}
+    }
 
-static void ServerListFetchServersCheck(WindowBase* w)
-{
-    if (_fetchFuture.valid())
+    void ServerListFetchServersBegin()
     {
-        auto status = _fetchFuture.wait_for(std::chrono::seconds::zero());
-        if (status == std::future_status::ready)
+        if (_fetchFuture.valid())
         {
+            // A fetch is already in progress
+            return;
+        }
+
+        _serverList.Clear();
+        _serverList.ReadAndAddFavourites();
+        _statusText = STR_SERVER_LIST_CONNECTING;
+
+        _fetchFuture = std::async(std::launch::async, [this] {
+            // Spin off background fetches
+            auto lanF = _serverList.FetchLocalServerListAsync();
+            auto wanF = _serverList.FetchOnlineServerListAsync();
+
+            // Merge or deal with errors
+            std::vector<ServerListEntry> allEntries;
             try
             {
-                auto [entries, statusText] = _fetchFuture.get();
-                _serverList.AddOrUpdateRange(entries);
-                _serverList.WriteFavourites(); // Update favourites in case favourited server info changes
-                _numPlayersOnline = _serverList.GetTotalPlayerCount();
-                _statusText = STR_X_PLAYERS_ONLINE;
-                if (statusText != STR_NONE)
-                {
-                    _statusText = statusText;
-                }
+                auto entries = lanF.get();
+                allEntries.insert(allEntries.end(), entries.begin(), entries.end());
+            }
+            catch (...)
+            {
+            }
```

Comment:
```suggestion
            // TODO: Stop catching all exceptions
            catch (...)
            {
            }
```

Reply:
For this one and the other `catch (...)` block, I think if there is an error it will already be caught by the underlying async call happening in the `try` block (i.e., `FetchLocalServerListAsync`. But it definitely won't hurt to be more specific in here too, so will mirror those exceptions in the general catch blocks in this class!

Reply:
Note that I'm not asking for you to fix this catch by narrowing it, just to add a Todo so we can track future improvement work. I think reworking it is our of the scope of this PR

Reply:
Oh ok perf, in that case I will just add the Todo! Thanks!